proposal/color-4-new-spaces-js.d.ts.md

```diff
@@ -218,68 +285,190 @@ colors as best they can, but it may be necessary in some cases.
-
-Returns a new SassColor as the result of changing some of [`internal`]'s
-channels. The `space` value defaults to the `space` of [`internal`], and any
-combination of channels in that space may be changed. Throws an error if any
-`channel` is not present in `space`.
-
-* If `space` is not defined, let `space` be the value of [`space(internal)`].
+#### `change`
 
-* Let `arguments` be the key value pairs in `channels` as keyword arguments.
+Replace the definition of [color.change] with the following:
+
+[color.change]: ../spec/js-api/value/color.d.ts.md#change
 
-* Return the value of the result of [`change(internal, ...arguments)`].
-
-[`space(internal)`]: #space
-
-[`change(internal, ...arguments)`]: ./color-4-new-spaces.md#colorchange
+This algorithm takes a JavaScript object `options` and returns a new SassColor
+as the result of changing some of [`internal`]'s components.
+
+The `space` value defaults to the `space` of [`internal`], and any combination
+of channels and alpha in that space may be changed.
+
+If `space` is not a [legacy color space], a channel value of `null` will result
+in a [missing component][missing components] value for that channel.
+
+* Let `initialSpace` be the value of [`internal.space()`].
+
+* If `options.space` is not defined, let `space` be the value of `initialSpace`.
+
+* Let `keys` be a list of the keys in `options` without `space`.
+
+* Let `components` be `"alpha"` and the channels in `space`.
+
+* If any key in `keys` is not the name of a channel in `components`, throw an
+  error.
+
+* If `space` is not equal to `initialSpace`, let `convertedColor` be the result
+  of [`internal.toSpace(space)`].
+
+* If `space` equals `hsl`, let `changedColor` be the result of:
+
+```js
+  SassColor({
+    hue: options.hue ?? internal.channel('hue'),
+    saturation: options.saturation ?? internal.channel('saturation'),
+    lightness: options.lightness ?? internal.channel('lightness'),
+    alpha: options.alpha ?? internal.channel('alpha'),
+    space: space
+  })
+  ```
+
+* If space equals `hwb`, let `changedColor` be the result of:
+
+```js
+  SassColor({
+    hue: options.hue ?? internal.channel('hue'),
+    whiteness: options.whiteness ?? internal.channel('whiteness'),
+    blackness: options.blackness ?? internal.channel('blackness'),
+    alpha: options.alpha ?? internal.channel('alpha'),
+    space: space
+  })
+  ```
+
+* If space equals `rgb`, let `changedColor` be the result of:
+
+```js
+  SassColor({
+    red: options.red ?? internal.channel('red'),
+    green: options.green ?? internal.channel('green'),
+    blue: options.blue ?? internal.channel('blue'),
+    alpha: options.alpha ?? internal.channel('alpha'),
+    space: space
+  })
+  ```
+
+* If space equals `lab` or `oklab`, let `changedColor` be the result of:
+
+  ```js
+  SassColor({
+    lightness: keys.includes('lightness') ? options.lightness : internal.channel('lightness'),
+    a: keys.includes('a') ? options.a : internal.channel('a'),
```

Comment:
Since we're changing `channel()` not to return `null`, this will need a little more logic to handle already-missing channel values. You could potentially factor that out into a procedure.

Reply:
I refactored this out into a procedure in 4d94520cc6925c432515ee886625ed57294486d9